auth/mux.go

```diff
@@ -0,0 +1,132 @@
+// Copyright 2016 The Serviced Authors.
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package auth
+
+import (
+	"bytes"
+	"encoding/binary"
+	"errors"
+)
+
+/*
+   When establishing a connection to the mux, in addition to the address of the receiver,
+   the sender sends an authentication token and signs the whole message. The token determines
+   if the sender is authorized to send data to the receiver or not
+
+   ----------------------------------------------------------------------------------------------
+   | Auth Token length (4 bytes)  |     Auth Token (N bytes)  | Address (6 bytes) |  Signature  |
+   ----------------------------------------------------------------------------------------------
+*/
+
+var (
+	endian           = binary.BigEndian
+	ErrBadMuxAddress = errors.New("Bad mux address")
+	ErrBadMuxHeader  = errors.New("Bad mux header")
+)
+
+const (
+	ADDRESS_BYTES   = 6
+	TOKEN_LEN_BYTES = 4
+)
+
+func BuildMuxHeader(address []byte) ([]byte, error) {
+	headerBuf := new(bytes.Buffer)
+
+	if len(address) != ADDRESS_BYTES {
+		return nil, ErrBadMuxAddress
+	}
+
+	//get current host token
+	token := AuthToken()
+
+	// add token length
+	var tokenLen uint32 = uint32(len(token))
+	tokenLenBuf := make([]byte, 4)
+	endian.PutUint32(tokenLenBuf, tokenLen)
+	headerBuf.Write(tokenLenBuf)
+
+	// add token
+	headerBuf.Write([]byte(token))
+
+	// add address
+	headerBuf.Write([]byte(address))
+
+	// Sign what we have so far
+	myPrivateKey := LocalPrivateKey()
+	signer, err := RSASigner(myPrivateKey)
+	if err != nil {
+		return nil, err
+	}
+	signature, err := signer.Sign(headerBuf.Bytes())
+	if err != nil {
+		return nil, err
+	}
+	// add signature to header
+	headerBuf.Write(signature)
+
+	return headerBuf.Bytes(), nil
+}
+
+func errorExtractingHeader(err error) (string, Identity, error) {
+	return "", nil, err
+}
+
+func ExtractMuxHeader(rawHeader []byte) (string, Identity, error) {
```

Comment:
The first return type should be `[]byte`. The address isn't a string. It's packed. Four bytes for the IP and two for the port.